### PR TITLE
[SYCL][NFCI] Drop ESIMD emulator leftovers

### PR DESCRIPTION
**Pull request by AlexeySachkov**
intel/llvm#4020 introduced some kind of wrapper for SYCL kernels on host to be able to launch them on ESIMD emulator backend. Even though we had dropped that feature since then, we didn't remove corresponding `handler.hpp` modifications. This PR removes them.

The main effect expected from this PR is compilation time improvement: those kernel wrappers are specialized by kernel name, meaning that there will be plenty of extra useless functions emitted during host compilation pass for every kernel in a program.

This patch also uncovered a missing case in
`HostKernel::InstantiateKernelOnHost` which we couldn't ever encounter because we transformed a host kernel to always accept `nd_item`, thus always skipping problematic `item` code path.

This PR is not expected to introduce any functional changes, but since I'm not very familiar with the SYCL RT, I'm not entirely sure of that.

### CODE REVIEW

**Review by uditagarwal97 — COMMENTED** [review]
Changes LGTM overall. However, I'm not very familiar with how `InstantiateKernelOnHost` is used, so, I'd request an additional review from @aelovikov-intel

File: sycl/include/sycl/detail/helpers.hpp [hunk 1]
```diff
@@ -10,8 +10,8 @@
 
 #include <CL/__spirv/spirv_types.hpp> // for MemorySemanticsMask
 #include <sycl/access/access.hpp>     // for fence_space
-#include <sycl/detail/export.hpp>     // for __SYCL_EXPORT
-#include <sycl/memory_enums.hpp>      // for memory_order
+#include <sycl/detail/export.hpp> // for __SYCL_EXPORT
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/helpers.hpp:13]: Unnecessary?

File: sycl/include/sycl/handler.hpp [hunk 2]
```diff
@@ -918,8 +794,8 @@ class __SYCL_EXPORT handler {
         detail::KernelLambdaHasKernelHandlerArgT<KernelType,
                                                  LambdaArgType>::value;
 
-    KernelType *KernelPtr =
-        ResetHostKernel<KernelType, LambdaArgType, Dims>(KernelFunc);
+    MHostKernel.reset(
+        new detail::HostKernel<KernelType, LambdaArgType, Dims>(KernelFunc));
```
Comment (aelovikov-intel) [inline sycl/include/sycl/handler.hpp:798]: should it be `= std::make_unique`?

**Review by aelovikov-intel — APPROVED** [review]
> However, I'm not very familiar with how InstantiateKernelOnHost is used

Me neither :)

File: sycl/include/sycl/detail/helpers.hpp (hunk 1 above)
Comment (AlexeySachkov) [reply sycl/include/sycl/detail/helpers.hpp]: Indeed, reverted those changes in cf21174495433bc5d0f061bea5cf1f926859d7b1

File: sycl/include/sycl/handler.hpp (hunk 2 above)
Comment (AlexeySachkov) [reply sycl/include/sycl/handler.hpp]: Originally I simply reverted this code to its older state, but applied `make_unique` in ed3589e3300cfaccbf16f56c9488a41113961d33

**Review by AlexeySachkov — COMMENTED** [review]
@steffenlarsen should be able to share some insights about that method of `HostKernel`, because he was the one who [re-?]introduced it not that long ago

**Comment by steffenlarsen** [comment]
InstantiateKernelOnHost is a little odd, I agree. It should never execute on host, but the intention of it is to instantiate kernels on the host so that they are not optimized out in host code. This preserves debugging information for the kernels.

**Pull request merged**